### ISSUE TITLE
Training forward pass doesn't match inference architecture — BridgingModule replaces decoder

### ISSUE THREAD

**Issue by NeuralNoble** [davidbrowne17/csm-streaming#50]
Hey, thanks for putting this repo together — been using it for voice cloning and ran into some questions about the training architecture.

  ---

  ### 1. BridgingModule vs Decoder mismatch

  During training in `forward_and_loss()` (line ~849), codebooks 1-31 are predicted through:

  ```python
  decoder_states = bridging_module(backbone_out)  # Linear(2048→1024), randomly initialized
  logits_i = decoder_states @ model.audio_head[i-1]
```
  But during inference in Model.generate_frame() (models.py), codebooks 1-31 go through:

```python
  decoder_h = self.decoder(self.projection(curr_h), ...)  # 4-layer transformer, 100M params
  ci_logits = torch.mm(decoder_h[:, -1, :], self.audio_head[i - 1])
```

  The training path uses a single randomly initialized linear layer where inference uses the full pretrained decoder transformer. The BridgingModule is never used at inference, and
   the decoder is never called during training.

  Questions:
  - Was this an intentional design choice for training efficiency?
  - Has anyone tried using the actual decoder during training (teacher forcing with ground truth codebook 0)? Curious if that improves audio quality for codebooks 1-31.

  ---
  2. Decoder LoRA weights get zero gradients

  replace_linear_with_lora() is called on the full model (line ~489), which applies LoRA to both the backbone and the decoder's q_proj, k_proj, v_proj, output_proj, w1, w2, w3
  layers.

  However since forward_and_loss() never calls model.decoder, those decoder LoRA params sit in the optimizer but receive zero gradients every step. They're allocated, tracked, but
  never learn anything (~3M wasted params).

  Is this expected, or should LoRA only be applied to the backbone?

  ---
  3. loadandmergecheckpoint.py — target layers don't match training

  The merge script has:

  # loadandmergecheckpoint.py
  target_layers = ['q_proj', 'k_proj', 'v_proj', 'o_proj']
  R=32
  APLHA=32

  But training in lora.py uses:

  # lora.py
  target_layers = ["q_proj", "k_proj", "v_proj", "output_proj", "w1", "w2", "w3"]
  R=32
  APLHA=32

  Two issues:
  - o_proj should be output_proj (the layer name in the model)
  - w1, w2, w3 (FFN layers) are trained but not loaded during merge

  This means loadandmergecheckpoint.py creates a LoRA structure that doesn't match the checkpoint, so load_state_dict(strict=False) silently skips most trained weights. The merged
  model is essentially the base model.

  Also R and APLHA are hardcoded instead of imported from lora.py — if someone changes them during training, the merge will silently corrupt weights.

  ---
  4. Validation loss samples from training set

  calculate_validation_loss() (line ~687) samples randomly from the same dataset used for training:

  val_indices = torch.randperm(len(dataset))[:max_samples].tolist()
  val_samples = [dataset[i] for i in val_indices]

  There's no train/val split, so validation loss always tracks training loss and can't detect overfitting.

  ---
  5. single_pass_forward is unused

  single_pass_forward() (line ~738) is defined but never called in the training loop — forward_and_loss() is used instead. Interestingly, single_pass_forward actually uses
  model.decoder, which would be closer to the inference path. Was this the original training function that got replaced by forward_and_loss?

  ---
  Would love to understand the reasoning behind the BridgingModule approach — maybe there's context I'm missing. Happy to contribute a PR for the loadandmergecheckpoint.py fixes at
   minimum. Great repo otherwise, the streaming inference is really solid.
